charms/reactive/endpoints.py

```diff
@@ -513,7 +513,7 @@ def received_app(self):
         if self._remote_app_data is None:
             # using JSONUnitDataView though it's name includes unit.
             self._remote_app_data = JSONUnitDataView(hookenv.relation_get(
-                app=self.application_name,
+                app=hookenv.application_name(),
```

Comment:
OK so this semantically changes the behaviour of this code which is why the original unit tests fail.

hookenv.application_name() always returns the local application name.

self.application_name in this context relates to the remote application name - so this change would work OK for a peer relation but not for a standard provides or requires relation

Reply:
```
    @property
    def application_name(self):
        """
        The name of the remote application for this relation, or ``None``.

        This is equivalent to::
        
            relation.units[0].unit_name.split('/')[0]
        """
        if self._application_name is None and self.units:
            self._application_name = self.units[0].unit_name.split('/')[0]
        return self._application_name
```
for reference